File: IPWorks Auth Samples/SAML/java/saml.java
/*
 * IPWorks Auth 2022 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks Auth in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksauth
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import ipworksauth.*;
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class saml extends ConsoleDemo {
	public static void main(String[] args) {
		Saml saml = new Saml();

		try {

			saml
					.addSamlEventListener(new ipworksauth.DefaultSamlEventListener() {
						public void SSLServerAuthentication(
								ipworksauth.SamlSSLServerAuthenticationEvent e) {
							// For the purposes of this demo we will accept any
							// server certificate
							e.accept = true;
						}
					});
			String authMode = "";
			while (!authMode.equalsIgnoreCase("q")) {
				System.out
						.println("******************************************************************************************");
				System.out
						.println("To begin, select the AuthMode. For SharePoint Online and Dynamics CRM simply specify your \r\n"
								+ "email address, password, and ApplicationURN. For ADFS you must also specify a LocalSTS. Note that \r\n"
								+ "this demo is configured for simple common scenarios. Please see the help file for more details.");
				System.out
						.println("******************************************************************************************");

				System.out.println("Please select an AuthMode:");
				System.out.println("1)  SharePoint Online");
				System.out.println("2)  Dynamics CRM");
				System.out.println("3)  ADFS");
				System.out.println("Q)  Quit");

				authMode = prompt("\r\noption", ">");

				// Initial setup
				if (authMode.equals("1")) // SharePoint Online
				{
					saml.setAuthMode(Saml.camSharePointOnline);
					saml.setUser(prompt("Username", ":",
							"user@mycrm.onmicrosoft.com"));
					saml.setPassword(prompt("Password"));
					saml.setApplicationURN(prompt("Application URN", ":",
							"https://mycrm.sharepoint.com"));
				} else if (authMode.equals("2")) {
					saml.setAuthMode(Saml.camDynamicsCRM);
					saml.setUser(prompt("Username", ":",
							"user@mycrm.onmicrosoft.com"));
					saml.setPassword(prompt("Password"));
					saml.setApplicationURN(prompt("Application URN", ":",
							"urn:crmapac:dynamics.com"));
				} else if (authMode.equals("3")) {
					saml.setAuthMode(Saml.camADFS);
					saml.setUser(prompt("User"));
					saml.setPassword(prompt("Password"));
					saml
							.setApplicationURN(prompt("Application URN", ":",
									"https://fsweb.contoso.com/ClaimsAwareWebAppWithManagedSTS/"));
					saml
							.setLocalSTS(prompt("Local STS", ":",
									"https://adfs.contoso.com/adfs/services/trust/2005/usernamemixed"));
				} else if (!authMode.equalsIgnoreCase("q")){
					System.out.println("Error: Input not recognized.");
					continue;
				} else {
					continue;
				}

				// If using SharePoint Online or Dynamics CRM call
				// GetSecurityToken()
				if (saml.getAuthMode() == Saml.camSharePointOnline
						|| saml.getAuthMode() == Saml.camDynamicsCRM) {
					System.out.println("\r\nObtaining Security Token ...");
					saml.getSecurityToken();
					System.out.println("\r\nSecurity Token:\r\n"
							+ saml.getSecurityTokenXML());
				} else // If using ADFS call GetAssertion to get the assertion
						// from the LocalSTS
				{
					System.out.println("\r\nObtaining Assertion ...");
					saml.getAssertion();
					System.out.println("\r\nAssertion:\r\n"
							+ saml.getAssertionXML());
				}
			}

		} catch (IPWorksAuthException ex) {
			System.out.println("IPWorksAuth exception thrown: " + ex.getCode()
					+ " [" + ex.getMessage() + "].");
		} catch (Exception ex) {
			System.out.println(ex.getMessage());
		}
	}
}
class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }

  static String prompt(String label, String punctuation, String defaultVal)
  {
	System.out.print(label + " [" + defaultVal + "] " + punctuation + " ");
	String response = input();
	if(response.equals(""))
		return defaultVal;
	else
		return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof IPWorksAuthException) {
      System.out.print(" (" + ((IPWorksAuthException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }
}

File: IPWorks Auth Samples/SAML/c++/saml.cpp
/*
 * IPWorks Auth 2022 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks Auth in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksauth
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */


#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/ipworksauth.h"
#define LINE_LEN 80

class MySAML : public SAML
{
public:
  virtual int FireSSLServerAuthentication (SAMLSSLServerAuthenticationEventParams *e)
  {
    e->Accept = true;
    return 0;
  }
};

int main(int argc, char **argv)
{
	MySAML saml;
	char buffer[LINE_LEN];
	buffer[strlen(buffer)-1] = '\0';
	int ret_code=0;

while((strcmp(buffer, "q") != 0) && (strcmp(buffer, "Q") != 0))
{
  printf("Please select an AuthMode:\n");
  printf("1) SharePoint Online\n");
  printf("2) Dynamics CRM\n");
  printf("3) ADFS\n");
  printf("Q) Quit\n");
  printf("Option> ");

	fgets(buffer, LINE_LEN, stdin);
  buffer[strlen(buffer)-1] = '\0';
  
  if (! strcmp(buffer, "1")) // SharePoint Online
  {
    saml.SetAuthMode(CAM_SHARE_POINT_ONLINE);
    saml.SetUser("user@mycrm.onmicrosoft.com");
    saml.SetApplicationURN("https://mycrm.sharepoint.com");

    printf("\nUsername (user@mycrm.onmicrosoft.com): ");
	  fgets(buffer, LINE_LEN, stdin);
	  buffer[strlen(buffer)-1] = '\0';
    if (strlen(buffer) > 0)
      saml.SetUser(buffer);

    printf("Password: ");
    fgets(buffer, LINE_LEN, stdin);
    buffer[strlen(buffer)-1] = '\0';
    saml.SetPassword(buffer);

    printf("Application URN (https://mycrm.sharepoint.com): ");
    fgets(buffer, LINE_LEN, stdin);
    buffer[strlen(buffer)-1] = '\0';
    if (strlen(buffer) > 0)
      saml.SetApplicationURN(buffer);

    printf("Obtaining Security Token...\n\n");
    ret_code = saml.GetSecurityToken();
    if (ret_code) goto done;

    printf("Security Token:\n%s\n\n", saml.GetSecurityTokenXML());
  } 
  else if (! strcmp(buffer, "2")) // Dynamics CRM
  {
    saml.SetAuthMode(CAM_DYNAMICS_CRM);
    saml.SetUser("user@mycrm.onmicrosoft.com");
    saml.SetApplicationURN("urn:crmapac:dynamics.com");

    printf("\nUsername (user@mycrm.onmicrosoft.com): ");
	  fgets(buffer, LINE_LEN, stdin);
	  buffer[strlen(buffer)-1] = '\0';
	  if (strlen(buffer) > 0)
      saml.SetUser(buffer);

    printf("Password: ");
    fgets(buffer, LINE_LEN, stdin);
    buffer[strlen(buffer)-1] = '\0';
    saml.SetPassword(buffer);

    printf("Application URN (urn:crmapac:dynamics.com): ");
    fgets(buffer, LINE_LEN, stdin);
    buffer[strlen(buffer)-1] = '\0';
    if (strlen(buffer) > 0)
      saml.SetApplicationURN(buffer);

    printf("Obtaining Security Token...\n\n");
    ret_code = saml.GetSecurityToken();
    if (ret_code) goto done;

    printf("Security Token:\n%s\n\n", saml.GetSecurityTokenXML());
  } 
  else if (! strcmp(buffer, "3")) // ADFS
  {
    saml.SetAuthMode(CAM_ADFS);
    saml.SetApplicationURN("https://fsweb.contoso.com/ClaimsAwareWebAppWithManagedSTS/");
    saml.SetLocalSTS("https://adfs.contoso.com/adfs/services/trust/2005/usernamemixed");

    printf("\nUser: ");
	  fgets(buffer, LINE_LEN, stdin);
	  buffer[strlen(buffer)-1] = '\0';
	  saml.SetUser(buffer);

    printf("Password: ");
    fgets(buffer, LINE_LEN, stdin);
    buffer[strlen(buffer)-1] = '\0';
    saml.SetPassword(buffer);

    printf("Application URN (https://fsweb.contoso.com/ClaimsAwareWebAppWithManagedSTS/): ");
    fgets(buffer, LINE_LEN, stdin);
    buffer[strlen(buffer)-1] = '\0';
    if (strlen(buffer) > 0)
      saml.SetApplicationURN(buffer);

    printf("Local STS (https://adfs.contoso.com/adfs/services/trust/2005/usernamemixed): ");
    fgets(buffer, LINE_LEN, stdin);
    buffer[strlen(buffer)-1] = '\0';
    if (strlen(buffer) > 0)
      saml.SetLocalSTS(buffer);
    
    printf("Obtaining Assertion...\n\n");
    ret_code = saml.GetAssertion();
    if (ret_code) goto done;

    printf("Assertion:\n%s\n\n", saml.GetAssertionXML());
  }
  else if(strcmp(buffer, "q") && strcmp(buffer, "Q"))
  {
    printf("\nInvalid Option\n\n");
  }
  else
  {
    continue;
  }
  
done:
	if (ret_code)     // Got an error.  The user is done.
	{
		printf( "Error: %d", ret_code );
    if (saml.GetLastError())
		{
			printf( " \"%s\"\n\n", saml.GetLastError() );
		}
	}

  fprintf(stderr, "press <return> to continue...\n");
	getchar();
}
	exit(ret_code);
	return 0;

}











